daemon/graphdriver/quota.go

```diff
@@ -0,0 +1,284 @@
+		}
+		fsx.fsx_projid = C.__u32(projectID)
+		fsx.fsx_xflags |= C.FS_XFLAG_PROJINHERIT
+		_, _, errno = syscall.Syscall(syscall.SYS_IOCTL, getDirFd(dir), C.FS_IOC_FSSETXATTR,
+			uintptr(unsafe.Pointer(&fsx)))
+		if errno != 0 {
+			return fmt.Errorf("Failed to set projid for %s: %v", targetPath, errno.Error())
+		}
+
+		q.quotas[targetPath] = projectID
+		q.nextProjectID++
+	}
+
+	//
+	// set the quota limit for the container's project id
+	//
+	var d C.fs_disk_quota_t
+	d.d_version = C.FS_DQUOT_VERSION
+	d.d_id = fsx.fsx_projid
+	d.d_flags = C.XFS_PROJ_QUOTA
+
+	d.d_fieldmask = C.FS_DQ_BHARD
+	d.d_blk_hardlimit = C.__u64(quota.Size / 512)
+
+	var cs = C.CString(q.backingFsDev)
+	defer C.free(unsafe.Pointer(cs))
+
+	_, _, errno := syscall.Syscall6(syscall.SYS_QUOTACTL, C.Q_XSETPQLIM,
+		uintptr(unsafe.Pointer(cs)), uintptr(fsx.fsx_projid),
+		uintptr(unsafe.Pointer(&d)), 0, 0)
+	if errno != 0 {
+		return fmt.Errorf("Failed to set quota limit for projid %d on %s: %v",
+			projectID, q.backingFsDev, errno.Error())
+	}
+
+	d.d_fieldmask = C.FS_DQ_BSOFT
+	d.d_blk_softlimit = C.__u64(quota.Size / 512)
+
+	_, _, errno = syscall.Syscall6(syscall.SYS_QUOTACTL, C.Q_XSETPQLIM,
+		uintptr(unsafe.Pointer(cs)), uintptr(fsx.fsx_projid),
+		uintptr(unsafe.Pointer(&d)), 0, 0)
+	if errno != 0 {
+		return fmt.Errorf("Failed to set quota limit for projid %d on %s: %v",
+			projectID, q.backingFsDev, errno.Error())
+	}
+
+	return nil
+}
+
+func (q *QuotaCtl) GetQuota(targetPath string, quota *Quota) error {
+
+	projectID, ok := q.quotas[targetPath]
+	if !ok {
+		return fmt.Errorf("quota not found for path : %s", targetPath)
+	}
+
+	//
+	// get the quota limit for the container's project id
+	//
+	var d C.fs_disk_quota_t
+
+	var cs = C.CString(q.backingFsDev)
+	defer C.free(unsafe.Pointer(cs))
+
+	_, _, errno := syscall.Syscall6(syscall.SYS_QUOTACTL, C.Q_XGETPQUOTA,
+		uintptr(unsafe.Pointer(cs)), uintptr(C.__u32(projectID)),
+		uintptr(unsafe.Pointer(&d)), 0, 0)
+	if errno != 0 {
+		return fmt.Errorf("Failed to get quota limit for projid %d on %s: %v",
+			projectID, q.backingFsDev, errno.Error())
+	}
+	quota.Size = uint64(d.d_blk_hardlimit) * 512
+
+	return nil
+}
+
+func getMaxProjectId(path string) (uint32, error) {
+
+	var projectID uint32 = 1
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		return 0, fmt.Errorf("read directory failed : %s", path)
+	}
+	for _, file := range files {
+		if !file.IsDir() {
+			continue;
+		}
+		dir, err := openDir(filepath.Join(path, file.Name()))
+		if err != nil {
+			fmt.Errorf("open dir failed %s", filepath.Join(path, file.Name()))
+			return 0, err
+		}
+		defer closeDir(dir)
+		var fsx C.struct_fsxattr
+		_, _, errno := syscall.Syscall(syscall.SYS_IOCTL, getDirFd(dir), C.FS_IOC_FSGETXATTR,
+			uintptr(unsafe.Pointer(&fsx)))
+		if errno != 0 {
+			continue
+		}
+		if projectID <= uint32(fsx.fsx_projid) {
+			projectID = uint32(fsx.fsx_projid) + 1
+		}
```

Comment:
don't you want to set this projid in the q.quotas[] map? if you don't, users won't be able to call getQuota() for those directories later on